### PR TITLE
Add rudimentary sprite patching support

### PR DESCRIPTION
**Pull request by jroweboy**
Adds a mesia sprite selectable sprite option before randomization.

What this change includes:
[x] Mapping from simplified character sprite sheet to CHR ROM
[x] Mesia sprite sheet to replace Semia

What this does not include:
[ ] Palette swapping. Only CHR ROM is edited atm.
[ ] Other characters changed, only Semia right now.
[ ] Complete graphics pack solution (some ground work is laid out)

Heres what it looks like:

image

### CODE REVIEW

File: scripts/update.sh [hunk 1]
```diff
@@ -34,8 +34,8 @@ done
 
 # Pick a default rom if present
 if [ -z "$rom" ]; then
-  rom=$(crc32 *.nes | grep ^1bd39032 | head -1)
-  rom=$(echo ${rom#1bd39032})
+  rom=$(sha256sum *.nes | grep ^070d22fe | head -1)
```
Comment (jroweboy) [inline scripts/update.sh:37]: crc32 isn't default on many distros, so I swapped to sha256sum which is more common

File: src/js/cli.ts [hunk 2]
```diff
@@ -131,7 +131,7 @@ const main = (...args: string[]) => {
     const filename = fillTemplate(output, {c, n, s, v, f, '%': '%'}) + '.nes';
     await new Promise(
         (resolve, reject) => fs.writeFile(
-            filename, shuffled, (err) => err ? reject(err) : resolve()));
+            filename, shuffled, (err) => err ? reject(err) : resolve('')));
```
Comment (jroweboy) [inline src/js/cli.ts:134]: got sick of seeing this as an error in my editor of choice (even though it doesn't error out when compiled to js)

File: src/js/characters.ts [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+  private generateSemiaMapping() : Map<number, number[]> {
+    // For most of the mappings, there is only one location to write it to, but for some, its split across several CHRROM banks.
+    // so thats why its a map of tileset number to a list of addresses
+
+    // A CHR tileset is 16 tiles wide, the left 8 tiles are the no armor sprites, the right 8 are the armor sprites
+    const ARMOR_TILESET_OFFSET = 0x08;
+    // In the original game, the game swaps the two 0x20 size CHR banks for sprites to switch armor/no armor sprites
+    const CHR_PAGE_OFFSET = 0x400;
+
+    const mapping = new Map<number, number[]>();
```
Comment (shicks) [inline src/js/characters.ts:73]: I might prefer to see this as a constructor argument, or some other format.  But this is fine for now.

File: src/js/characters.ts [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+const mesia_patch_data = Array.from(atob(mesia_spritesheet_chr), c => c.charCodeAt(0))
```
Comment (shicks) [inline src/js/characters.ts:378]: atob is deprecated, TextEncoder/TextDecoder is the preferred solution now.  Hopefully this gets cleaned up, tho, so I don't really care too much - that's mainly FYI.

**Review by shicks — COMMENTED** [review]
I'm going to pull this into a `sprite` branch.

**Comment by shicks** [comment]
Merged in 84b3a94

**Pull request closed without merging**